com.unity.ml-agents/Runtime/Actuators/ActionSegment.cs

```diff
@@ -0,0 +1,181 @@
+        {
+            if (offset + length > actionArray.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset),
+                    $"Arguments offset: {offset} and length: {length} " +
+                    $"are out of bounds of actionArray: {actionArray.Length}.");
+            }
+        }
+#endif
+
+        /// <summary>
+        /// Construct an <see cref="ActionSegment{T}"/> with an underlying array
+        /// and offset, and a length.
+        /// </summary>
+        /// <param name="actionArray">The underlying array which this segment has a view into</param>
+        /// <param name="offset">The zero-based offset into the underlying array.</param>
+        /// <param name="length">The length of the segment.</param>
+        public ActionSegment(T[] actionArray, int offset, int length)
+        {
+            CheckParameters(actionArray, offset, length);
+            Array = actionArray;
+            Offset = offset;
+            Length = length;
+        }
+
+        /// <summary>
+        /// Get the underlying <see cref="Array"/> of this segment.
+        /// </summary>
+        public T[] Array { get; }
+
+        /// <summary>
+        /// Allows access to the underlying array using array syntax.
+        /// </summary>
+        /// <param name="index">The zero-based index of the segment.</param>
+        /// <exception cref="IndexOutOfRangeException">Thrown when the index is less than 0 or
+        /// greater than or equal to <see cref="Length"/></exception>
+        public T this[int index]
+        {
+            get
+            {
+                if (index < 0 || index > Length)
+                {
+                    throw new IndexOutOfRangeException($"Index out of bounds, expected a number between 0 and {Length}");
+                }
+                return Array[Offset + index];
+            }
+        }
+
+        /// <inheritdoc cref="IEnumerable{T}.GetEnumerator"/>
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        /// <inheritdoc cref="IEnumerable{T}"/>
+        public IEnumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        /// <inheritdoc cref="ValueType.Equals(object)"/>
+        public override bool Equals(object obj)
+        {
+            if (!(obj is ActionSegment<T>))
+            {
+                return false;
+            }
+            return Equals((ActionSegment<T>)obj);
+        }
+
+        /// <inheritdoc cref="IEquatable{T}.Equals(T)"/>
+        public bool Equals(ActionSegment<T> other)
+        {
+            return Offset == other.Offset && Length == other.Length && Equals(Array, other.Array);
+        }
+
+        /// <inheritdoc cref="ValueType.GetHashCode"/>
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                var hashCode = Offset;
```

Comment:
Can we use [HashCode.Combine()](https://docs.microsoft.com/en-us/dotnet/api/system.hashcode.combine?view=dotnet-plat-ext-3.1) or is that too new?